modules/light-clients/07-celestia/light_client_module.go

```diff
@@ -0,0 +1,237 @@
+	clientStore := l.storeProvider.ClientStore(ctx, clientID)
+
+	clientState, found := getClientState(clientStore, l.cdc)
+	if !found {
+		return clienttypes.ZeroHeight()
+	}
+
+	return clientState.BaseClient.LatestHeight
+}
+
+// TimestampAtHeight obtains the client state associated with the client identifier and calls into the clientState.GetTimestampAtHeight method.
+//
+// CONTRACT: clientID is validated in 02-client router, thus clientID is assumed here to have the format 07-celestia-{n}.
+func (l LightClientModule) TimestampAtHeight(
+	ctx sdk.Context,
+	clientID string,
+	height exported.Height,
+) (uint64, error) {
+	clientStore := l.storeProvider.ClientStore(ctx, clientID)
+	clientState, found := getClientState(clientStore, l.cdc)
+	if !found {
+		return 0, errorsmod.Wrap(clienttypes.ErrClientNotFound, clientID)
+	}
+
+	return clientState.BaseClient.GetTimestampAtHeight(ctx, clientStore, l.cdc, height)
+}
+
+// RecoverClient asserts that the substitute client is a celestia client. It obtains the client state associated with the
+// subject client and calls into the subjectClientState.CheckSubstituteAndUpdateState method.
+//
+// CONTRACT: clientID is validated in 02-client router, thus clientID is assumed here to have the format 07-celestia-{n}.
+func (l LightClientModule) RecoverClient(ctx sdk.Context, clientID, substituteClientID string) error {
+	substituteClientType, _, err := clienttypes.ParseClientIdentifier(substituteClientID)
+	if err != nil {
+		return err
+	}
+
+	if substituteClientType != ModuleName {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "expected: %s, got: %s", ModuleName, substituteClientType)
+	}
+
+	clientStore := l.storeProvider.ClientStore(ctx, clientID)
+	clientState, found := getClientState(clientStore, l.cdc)
+	if !found {
+		return errorsmod.Wrap(clienttypes.ErrClientNotFound, clientID)
+	}
+
+	substituteClientStore := l.storeProvider.ClientStore(ctx, substituteClientID)
+	substituteClient, found := getClientState(substituteClientStore, l.cdc)
+	if !found {
+		return errorsmod.Wrap(clienttypes.ErrClientNotFound, substituteClientID)
+	}
+
+	return clientState.BaseClient.CheckSubstituteAndUpdateState(ctx, l.cdc, clientStore, substituteClientStore, substituteClient)
+}
+
+// VerifyUpgradeAndUpdateState obtains the client state associated with the client identifier and calls into the clientState.VerifyUpgradeAndUpdateState method.
+// The new client and consensus states will be unmarshaled and an error is returned if the new client state is not at a height greater
+// than the existing client.
+//
+// CONTRACT: clientID is validated in 02-client router, thus clientID is assumed here to have the format 07-celestia-{n}.
+func (l LightClientModule) VerifyUpgradeAndUpdateState(
+	ctx sdk.Context,
+	clientID string,
+	newClient []byte,
+	newConsState []byte,
+	upgradeClientProof,
+	upgradeConsensusStateProof []byte,
+) error {
+	// TODO: do we need to implement this?
```

Comment:
Question: do we need to implement this functionality?

Reply:
We can add an issue for this and discuss later with celestia team